compiler/rustc_mir_build/src/build/matches/mod.rs

```diff
@@ -780,23 +841,29 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         }
     }
 
+    /// Emits a [`StatementKind::StorageLive`] for the given var, and also
+    /// schedules a drop if requested (and possible).
     pub(crate) fn storage_live_binding(
         &mut self,
         block: BasicBlock,
         var: LocalVarId,
         span: Span,
         for_guard: ForGuard,
-        schedule_drop: bool,
+        schedule_drop: ScheduleDrops,
     ) -> Place<'tcx> {
         let local_id = self.var_local_id(var, for_guard);
         let source_info = self.source_info(span);
         self.cfg.push(block, Statement { source_info, kind: StatementKind::StorageLive(local_id) });
         // Although there is almost always scope for given variable in corner cases
         // like #92893 we might get variable with no scope.
-        if let Some(region_scope) = self.region_scope_tree.var_scope(var.0.local_id)
-            && schedule_drop
-        {
-            self.schedule_drop(span, region_scope, local_id, DropKind::Storage);
+        if let Some(region_scope) = self.region_scope_tree.var_scope(var.0.local_id) {
```

Comment:
Side note: I was looking at the history of this `if let` (#94849), and it made me wonder if the `None` case can never actually occur.

I tried replacing the `None` case with `span_bug` and couldn't observe any failures in the test suites I ran. That's not entirely conclusive, but it is suggestive.

(Obviously this is all outside the scope of *this* PR; just noting something interesting that I noticed.)

Reply:
Hm weird, that PR does introduce a test that used to get a `None`, so something else must have changed. I can't say I understand this code enough to tell though.

Reply:
My guess is that at the time the change was made, there were fewer “early” syntactic checks for `let` expressions in bad places, so it was possible for one to (illegally) exist inside certain const contexts that would survive to this point and cause an unwrap panic.

The change from unwrap to if-let was requested by https://github.com/rust-lang/rust/pull/94849#discussion_r833967008, but it doesn't give context for why. I suspect that it would have been preferable to make the None case do a `span_delayed_bug` instead of silently doing nothing.

Reply:
Looks like #115677 is what moved the syntactic checks to parse-time, guaranteeing that illegal let expressions don't proceed to MIR building. That gives me a bit more confidence that the None scenario can't happen, and that we have tests that would likely hit this edge-case if it were possible.